rclcpp/src/rclcpp/duration.cpp

```diff
@@ -69,24 +66,26 @@ Duration::Duration(const rcl_duration_t & duration)
 Duration::operator builtin_interfaces::msg::Duration() const
 {
   builtin_interfaces::msg::Duration msg_duration;
-  msg_duration.sec = static_cast<std::int32_t>(RCL_NS_TO_S(rcl_duration_.nanoseconds));
-  msg_duration.nanosec =
-    static_cast<std::uint32_t>(rcl_duration_.nanoseconds % (1000 * 1000 * 1000));
+  constexpr rcl_duration_value_t kRemainder = RCL_S_TO_NS(1);
+  const auto result = std::div(rcl_duration_.nanoseconds, kRemainder);
+  if (result.rem >= 0) {
+    msg_duration.sec = static_cast<std::int32_t>(result.quot);
+    msg_duration.nanosec = static_cast<std::uint32_t>(result.rem);
+  } else {
+    msg_duration.sec = static_cast<std::int32_t>(result.quot - 1);
+    msg_duration.nanosec = static_cast<std::uint32_t>(kRemainder + result.rem);
+  }
   return msg_duration;
 }
 
 Duration &
-Duration::operator=(const Duration & rhs)
-{
-  rcl_duration_.nanoseconds = rhs.rcl_duration_.nanoseconds;
-  return *this;
-}
+Duration::operator=(const Duration & rhs) = default;
 
 Duration &
 Duration::operator=(const builtin_interfaces::msg::Duration & duration_msg)
 {
-  rcl_duration_.nanoseconds = RCL_S_TO_NS(static_cast<int64_t>(duration_msg.sec));
-  rcl_duration_.nanoseconds += duration_msg.nanosec;
+  rcl_duration_.nanoseconds = RCL_S_TO_NS(static_cast<rcl_duration_value_t>(duration_msg.sec));
+  rcl_duration_.nanoseconds += static_cast<rcl_duration_value_t>(duration_msg.nanosec);
```

Comment:
to avoid duplication, you can take advantage of the constructor taking a `builtin_interfaces::msg::Duration` and then use copy assignment. In that way, we won't need to update this method again in the future.
I don't mind strongly though

Reply:
Fine for me. But unless the compiler optimizes that, which is probably what happens, constructing a temporary and invoking the copy assignment involves an additional step.

I will apply the same suggestion to [`Time::operator=(const builtin_interfaces::msg::Time & time_msg)`](https://github.com/ros2/rclcpp/blob/683ac64ee4edf52ae3a4cd1a722c17c4bd81b562/rclcpp/src/rclcpp/time.cpp#L110) for consistency.

Strictly speaking the overloads could be removed completely then, because the respective constructors are not marked `explicit` and the compiler will implicitly construct a temporary `Time` instance and invoke the default copy-assignment operator. But doing so would break the ABI compatibility.

Reply:
> Fine for me. But unless the compiler optimizes that, which is probably what happens, constructing a temporary and invoking the copy assignment involves an additional step.

That's a far point. I prefer deduplication over a bit of extra performance, but it might be arguable.